src/preprocessing/passes/bv_to_int.cpp

```diff
@@ -0,0 +1,981 @@
+  Node eliminated = d_eliminationCache[n];
+  Assert(d_rebuildCache.find(eliminated) != d_rebuildCache.end());
+  Assert(!d_rebuildCache[eliminated].isNull());
+  return d_rebuildCache[eliminated];
+}
+
+/**
+ * Translate n to Integers via post-order traversal.
+ */
+Node BVToInt::bvToInt(Node n)
+{
+  n = eliminationPass(n);
+  n = makeBinary(n);
+  vector<Node> toVisit;
+  toVisit.push_back(n);
+  uint64_t granularity = options::solveBVAsInt();
+
+  while (!toVisit.empty())
+  {
+    Node current = toVisit.back();
+    uint64_t currentNumChildren = current.getNumChildren();
+    if (d_bvToIntCache.find(current) == d_bvToIntCache.end())
+    {
+      // This is the first time we visit this node and it is not in the cache.
+      d_bvToIntCache[current] = Node();
+      toVisit.insert(toVisit.end(), current.begin(), current.end());
+    }
+    else
+    {
+      // We already visited this node
+      if (!d_bvToIntCache[current].isNull())
+      {
+        // We are done computing the translation for current
+        toVisit.pop_back();
+      }
+      else
+      {
+        // We are now visiting current on the way back up.
+        // This is when we do the actual translation.
+        if (currentNumChildren == 0)
+        {
+          if (current.isVar())
+          {
+            if (current.getType().isBitVector())
+            {
+              // For bit-vector variables, we create integer variables and add a
+              // range constraint.
+              Node newVar = d_nm->mkSkolem("__bvToInt_var",
+                                           d_nm->integerType(),
+                                           "Variable introduced in bvToInt "
+                                           "pass instead of original variable "
+                                               + current.toString());
+
+              d_bvToIntCache[current] = newVar;
+              d_rangeAssertions.insert(mkRangeConstraint(
+                  newVar, current.getType().getBitVectorSize()));
+            }
+            else
+            {
+              // Boolean variables are left unchanged.
+              AlwaysAssert(current.getType() == d_nm->booleanType()
+                           || current.getType().isSort());
+              d_bvToIntCache[current] = current;
+            }
+          }
+          else if (current.isConst())
+          {
+            if (current.getKind() == kind::CONST_BITVECTOR)
+            {
+              // Bit-vector constants are transformed into their integer value.
+              BitVector constant(current.getConst<BitVector>());
+              Integer c = constant.toInteger();
+              d_bvToIntCache[current] = d_nm->mkConst<Rational>(c);
+            }
+            else
+            {
+              // Other constants stay the same.
+              d_bvToIntCache[current] = current;
+            }
+          }
+          else
+          {
+            Unimplemented();
```

Comment:
Is this supposed to be a TODO? Or a case that can never happen?
If the latter, you should assert at the top that `current.isVar() || current.isConst()` and case split between those two. If it is a TODO, why is it not handled yet?

Reply:
The code now has only a single `Unimplemented()`, due to lack of support for logics that go beyond `QF_UFBV`, and this is now documented in the code. For this particular case that you commented on, I replaced it with the assertion that you suggested.